src/components/task-app/task-app.component.js

```diff
@@ -37,8 +37,33 @@ const TaskApp = () => {
     );
   };
 
+  const completeAllTasks = () => {
+    if (taskList.find((task) => !task.completed)) {
+      setTaskList(
+        taskList.map((task) => {
+          return {
+            ...task,
+            completed: true,
+            hoverMessage: "Mark as Incomplete",
+          };
+        })
+      );
+    } else {
+      setTaskList(
+        taskList.map((task) => {
+          return {
+            ...task,
+            completed: false,
+            hoverMessage: "Mark as Complete",
+          };
+        })
+      );
+    }
+  };
```

Comment:
With the complete all button I just did it like that because in the same button it's required to have the option to complete all task or to incomplete all of them. Being this the case does this still apply?

Reply:
if it is to complete/incomplete them all, `completeAllTasks` can take a boolean `done` that can be used instead of the `true` in the snippet above:

```typescript
  const completeAllTasks = (done) => {
    setTaskList(previous =>
      previous.map((task) => {
        return {
          ...task,
          completed: done
        };
      })
    );
  };
```

so, whenever you needed it, you can call it as `completeAllTasks(true)` to mark them **all** as completed, or `completeAllTasks(false)` to mark them **all** as incompleted.

Reply:
The only place that I'm calling this function is on a click button can this still be applied ?

Reply:
ohhh... so, if I get it right: when the user clicks on a button this function will check if there's at least one uncompleted task in the list. If so, it will change the state of all tasks to `true`, otherwise will make them all `false`.

If that's right, it'll be good to grab the value of the found task separately and then perform the operation. (very similar to what you have):

```javascript
  const completeAllTasks = () => {
    const uncompleted = taskList.find((task) => !task.completed);

    if (uncompleted) {
      setTaskList((prevState) =>
        prevState.map((task) => {
          return {
            ...task,
            completed: true,
          };
        })
      );
    } else {
      setTaskList((prevState) =>
        prevState.map((task) => {
          return {
            ...task,
            completed: false,
          };
        })
      );
    }
  };
```

you can then simplify this into:

```javascript
  const completeAllTasks = () => {
    const uncompleted = taskList.find((task) => !task.completed);

    setTaskList((prevState) =>
      prevState.map((task) => {
        return {
          ...task,
          completed: !!uncompleted,
        };
      })
    );
  };
```

since `uncompleted` (when found) is an object, we use the double exclamation mark `!!` to make it boolean (if the value exists it'll turn it into true).

Alternatively, you can make use of the `.some` array function:

```javascript
  const completeAllTasks = () => {
    const uncompleted = taskList.some((task) => !task.completed);

    setTaskList((prevState) =>
      prevState.map((task) => {
        return {
          ...task,
          completed: uncompleted,
        };
      })
    );
  };
```

In this case, `.some` will return a `boolean` if at least one of the elements in the array evaluates to true for the given predicate function.

Reply:
37a59ce2fa9b38d08cc692dda93ae4b417d7465e Changed as advised to simplify the function.